README.md
# pwa-studio README

This is the README for your extension "pwa-studio". After writing up a brief description, we recommend including the following sections.

## Features

Describe specific features of your extension including screenshots of your extension in action. Image paths are relative to this README file.

For example if there is an image subfolder under your extension project workspace:

\!\[feature X\]\(images/feature-x.png\)

> Tip: Many popular extensions utilize animations. This is an excellent way to show off your extension! We recommend short, focused animations that are easy to follow.

## Requirements

If you have any requirements or dependencies, add a section describing those and how to install and configure them.

## Extension Settings

Include if your extension adds any VS Code settings through the `contributes.configuration` extension point.

For example:

This extension contributes the following settings:

* `myExtension.enable`: enable/disable this extension
* `myExtension.thing`: set to `blah` to do something

## Known Issues

Calling out known issues can help limit users opening duplicate issues against your extension.

## Release Notes

Users appreciate release notes as you update your extension.

### 1.0.0

Initial release of ...

### 1.0.1

Fixed issue #.

### 1.1.0

Added features X, Y, and Z.

-----------------------------------------------------------------------------------------------------------
## Following extension guidelines

Ensure that you've read through the extensions guidelines and follow the best practices for creating your extension.

* [Extension Guidelines](https://code.visualstudio.com/api/references/extension-guidelines)

## Working with Markdown

**Note:** You can author your README using Visual Studio Code.  Here are some useful editor keyboard shortcuts:

* Split the editor (`Cmd+\` on macOS or `Ctrl+\` on Windows and Linux)
* Toggle preview (`Shift+CMD+V` on macOS or `Shift+Ctrl+V` on Windows and Linux)
* Press `Ctrl+Space` (Windows, Linux) or `Cmd+Space` (macOS) to see a list of Markdown snippets

### For more information

* [Visual Studio Code's Markdown Support](http://code.visualstudio.com/docs/languages/markdown)
* [Markdown Syntax Reference](https://help.github.com/articles/markdown-basics/)

**Enjoy!**


CHANGELOG.md
# Change Log

All notable changes to the "pwa-studio" extension will be documented in this file.

## [v1.0.4]
- New snippets for screenshots and shortcuts + remove pwa prefix from snippets for better discoverability: https://github.com/pwa-builder/pwa-studio/pull/75

## [v1.0.3]

- New Web Manifest editing and auditing experience: https://github.com/pwa-builder/pwa-studio/pull/68
    - Fixes https://github.com/pwa-builder/PWABuilder/issues/2671
    - Fixes https://github.com/pwa-builder/PWABuilder/issues/2395
    - Fixes https://github.com/pwa-builder/PWABuilder/issues/2403
    - Fixes https://github.com/pwa-builder/PWABuilder/issues/2629
    - Fixes https://github.com/pwa-builder/PWABuilder/issues/2400
    - Fixes https://github.com/pwa-builder/PWABuilder/issues/2384
- Extension is now bundled to reduce startup time: https://github.com/pwa-builder/pwa-studio/pull/67
- New snippets: https://github.com/pwa-builder/pwa-studio/pull/70
- New usage analytics: https://github.com/pwa-builder/pwa-studio/pull/72


## [v0.3.1]

- Fixed: https://github.com/pwa-builder/PWABuilder/issues/2635 

## [v0.3.0]

- Fixed: https://github.com/pwa-builder/PWABuilder/issues/2626
- Fixed: https://github.com/pwa-builder/PWABuilder/issues/2424
- New Help Panel feature: https://github.com/pwa-builder/PWABuilder/issues/2514
- PWA Starter now uses [Vite](https://github.com/pwa-builder/pwa-starter/wiki): https://github.com/pwa-builder/PWABuilder/issues/2670

## [v0.2.6]

- Workaround for analytics issue: https://github.com/pwa-builder/pwa-studio/compare/help-panel?expand=1#diff-3e1333fb5dae3594a3e5ba561f7937487c8fe8d3a1bba91c706e213b5b69e5b1L41
- Manifest and Icon generation views will not generate until required fields are filled out, including icon file input: https://github.com/pwa-builder/PWABuilder/issues/2627

## [v0.2.5]

- Added better usage analytics. No personal information is handled, only command usage. This will allow
us to better scope our work in the future to what is being used the most in PWA Studio.

## [v0.2.4]

- More readme updates for the VSCode Marketplace

## [v0.2.3]

- Updated readme for VSCode Marketplace


## [v0.2.2]

- Updated the Windows packaging URL to our new Windows packaging server

## [v0.2.1]

- Fixed: https://github.com/pwa-builder/PWABuilder/issues/2579
- Fixed: https://github.com/pwa-builder/PWABuilder/issues/2557

- Setup tests: https://github.com/pwa-builder/PWABuilder/issues/2588
- Better Typescript usage: https://github.com/pwa-builder/PWABuilder/issues/2595

## [v0.2.0]

- Better error messages
- Webviews now use the VSCode UI toolkit
- Submit buttons in webviews now have "generating..." UI to let the user know something is going on in the background
- We now use the built in VSCode icons
- Our Icon now matches the VSCode extension design guidelines
- Bug fixes around manifest generation
- Actions in the bottom bar now live in the view panels

## [v0.1.2]

- Paths for save dialogs are now correct
- Icons generated for the default manifest are converted to files
- diagnostics are only ran on JSON files

## [v0.1.0]

- Icons are now generated as files, not base64 strings.

## [v0.0.9]

- New in-manifest error prompts for key manifest issues.

## [v0.0.8]

- Major fix around windows paths causing the manifest / service worker to not be found.


src/views/icons-view.ts
import { writeFile } from "fs/promises";
import * as vscode from "vscode";
import { Manifest } from "../interfaces";
import {
  convertBaseToFile,
  findManifest,
} from "../services/manifest/manifest-service";
import { getUri } from "../utils";

import { getAnalyticsClient } from "../services/usage-analytics";

export class IconGenerationPanel {
  public static currentPanel: IconGenerationPanel | undefined;
  private readonly _panel: vscode.WebviewPanel;
  private _disposables: vscode.Disposable[] = [];

  private constructor(panel: vscode.WebviewPanel, extensionUri: vscode.Uri) {
    this._panel = panel;

    this._panel.onDidDispose(this.dispose, null, this._disposables);

    this._panel.webview.html = this._getWebviewContent(
      this._panel.webview,
      extensionUri
    );

    let iconsObject: any;
    // Handle messages from the webview
    this._panel.webview.onDidReceiveMessage(
      async (message) => {
        switch (message.command) {
          case "prompt":
            const analyticsClient = getAnalyticsClient();
            
            analyticsClient.trackEvent({ 
              name: "generate",  
              properties: { type: "icons"} 
            });

            iconsObject = message.iconsObject;
            const manifest: vscode.Uri = await findManifest();

            if (manifest && iconsObject.icons) {
              // read manifest file
              const manifestFile = await vscode.workspace.openTextDocument(
                manifest
              );

              const manifestObject: Manifest = JSON.parse(
                manifestFile.getText()
              );

              const newIconsData = await convertBaseToFile(iconsObject.icons);

              // add icons to manifest
              manifestObject.icons = newIconsData.icons;

              // write manifest file
              await writeFile(
                manifest.fsPath,
                JSON.stringify(manifestObject, null, 2)
              );

              // show manifest with vscode
              await vscode.window.showTextDocument(manifestFile);
            } else {
              vscode.window.showErrorMessage(
                "You first need a Web Manifest. Tap the Generate Manifest button at the bottom to get started."
              );
            }

            return;
        }
      },
      undefined,
      this._disposables
    );
  }

  private _getWebviewContent(
    webview: vscode.Webview,
    extensionUri: vscode.Uri
  ) {
    const toolkitUri = getUri(webview, extensionUri, [
      "node_modules",
      "@vscode",
      "webview-ui-toolkit",
      "dist",
      "toolkit.js",
    ]);

    // Tip: Install the es6-string-html VS Code extension to enable code highlighting below
    return /*html*/ `
    <!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <link rel="icon" href="https://glitch.com/favicon.ico" />

    <title>PWA VSCode Extension Manifest Form</title>

    <script type="module" src="${toolkitUri}"></script>

    <!-- Ionicons Import -->
    <script src="https://unpkg.com/ionicons@4.5.10-0/dist/ionicons.js"></script>
  </head>
  <body>
    <div id="central">
      <form id="manifest-options" onsubmit="handleSubmit(event)">

      <div id="submit-block">
        <h1>Generate Icons</h1>
        <button type="submit" id="submit" type="submit">Generate Icons</vscode-button>
      </div>

        <div id="first-six">
          <div class="six">
            <label for="file_input">Choose a 512x512 icon:</label>
            <div class="input-area">
              <input type="file" name="file_input" id="file_input" required />

              <a
                href="https://developer.mozilla.org/en-US/docs/Web/Manifest/icons"
                target="_blank"
                rel="noopener"
              >
                <ion-icon name="information-circle-outline"></ion-icon>
                <p class="toolTip">
                  Click for more info on icons your manifest.
                </p>
              </a>
            </div>
          </div>
        </div>
      </form>
    </div>
    <script>
      const vscode = acquireVsCodeApi();

      let file = undefined;
      document.querySelector("#file_input").addEventListener("change", (ev) => {
        file = ev.target.files[0];
      });

      async function generateIcons() {
        return new Promise(async (resolve, reject) => {

          const url =
            "https://appimagegenerator-prod.azurewebsites.net/api/image/base64";

          const form = new FormData();
          form.append("baseImage", file);
          form.append("platform", "windows10");
          form.append("platform", "android");
          form.append("platform", "ios");
          form.append("colorChanged", "false");
          form.append("padding", "0");
          // send formdata with http node module
          try {
            const response = await fetch(url, {
              method: "POST",
              body: form,
            });

            const data = await response.json();

            resolve(data);
          } catch (err) {
            console.error("error", err);
            reject(err);
          }
        });
      }

      async function handleSubmit(event) {
        event.preventDefault();
        // update button text
        document.querySelector("#submit").innerText = "Generating...";

        const icons = await generateIcons();
        
        document.querySelector("#submit").innerText = "Generate Icons";

        let maniObj = {
          icons: icons,
        };

        vscode.postMessage({
          command: "prompt",
          text: "Your icons have been generated!",
          iconsObject: maniObj,
        });

        event.preventDefault();
      }
    </script>
  </body>
  <style>
    body.vscode-light {
      color: black;
    }

    body.vscode-dark label {
      color: white;
    }

    #file_input {
      display: block;
      height: 2em;
      width: 14em;
    }

    #central {
      padding: 1em;
      font-family: sans-serif;
    }

    #manifest-options {
      display: flex;
      flex-direction: column;
    }

    .input-area {
      display: flex;
      justify-content: flex-start;
      align-items: center;
    }

    #first-six {
      display: grid;
      grid-template-columns: auto auto;
      grid-gap: 20px;
      margin: 20px 0;
    }

    .six {
      display: flex;
      flex-direction: column;
    }

    .toolTip {
      visibility: hidden;
      width: 160px;
      background-color: #f8f8f8;
      color: black;
      text-align: center;
      border-radius: 6px;
      padding: 5px;
      /* Position the tooltip */
      position: absolute;
      top: 0px;
      left: 70%;
      z-index: 1;
    }

    .input-area a {
      position: relative;
    }

    a:hover .toolTip {
      visibility: visible;
    }

    label {
      margin-bottom: 6px;
      font-size: 16px;
      color: black;

      font-weight: bold;
    }

    input {
      border-radius: 4px;
      box-sizing: border-box;
      border: 1px solid #a8a8a8;
      height: 38px;
      width: 95%;
      font-size: 14px;
      text-indent: 10px;
      color: black;
    }

    #file_input {
      border: none;
      color: currentColor;
    }

    select {
      border-radius: 4px;
      box-sizing: border-box;
      border: 1px solid #a8a8a8;
      height: 38px;
      width: 95%;
      font-size: 14px;
      text-indent: 10px;
    }

    textarea {
      margin-top: 6px;
      margin-bottom: 20px;
      border-radius: 4px;
      box-sizing: border-box;
      border: 1px solid #a8a8a8;
      height: 38px;
      width: 100%;
      font-size: 14px;
      text-indent: 10px;
      color: black;
      width: 45%;
    }

    #icon {
      padding: 1px;
    }

    #bottom-four {
      display: grid;
      grid-template-columns: 1fr 1fr;
      grid-gap: 20px;
    }

    .color {
      display: flex;
      flex-direction: column;
    }

    #icon {
      border: none;
      text-indent: 0;
      margin-top: 6px;
      height: max-content;
    }

    #bottom-four button {
      font-size: 16px;
      font-weight: bolder;
      padding: 20px 10px;

      border-radius: 30px;
      border: none;

      height: 75%;

      display: flex;
      align-items: center;
      justify-content: center;
    }

    #bottom-four button:hover {
      cursor: pointer;
    }

    #submit-block {
      display: flex;
      justify-content: space-between;
      align-items: center;
      margin-top: 2em;

      position: sticky;
      top: 0;
      z-index: 9;
      height: 100%;
      padding-top: 10px;

      border-bottom: solid 1px darkgrey;
      padding-bottom: 10px;
    }

    #submit-block h1 {
      font-size: 16px;
      margin-top: 0;
      margin-bottom: 0;
    }

    #submit-block button {
      background: #487cf1;
      color: white;
      border: none;
      padding: 12px;
      font-size: 1.1em;
      border-radius: 4px;
      cursor: pointer;
    }

    ion-icon {
      margin-left: 10px;
      font-size: 24px;
    }

    ion-icon:hover {
      cursor: pointer;
    }

    a:visited {
      color: black;
    }
  </style>
</html>

    `;
  }

  public static render(extensionUri: vscode.Uri) {
    if (IconGenerationPanel.currentPanel) {
      IconGenerationPanel.currentPanel._panel.reveal(vscode.ViewColumn.One);
    } else {
      const panel = vscode.window.createWebviewPanel(
        "iconsview",
        "Generate Icons",
        vscode.ViewColumn.Beside,
        {
          enableScripts: true,
        }
      );

      IconGenerationPanel.currentPanel = new IconGenerationPanel(
        panel,
        extensionUri
      );
    }
  }

  public dispose() {
    IconGenerationPanel.currentPanel = undefined;

    this._panel.dispose();

    while (this._disposables.length) {
      const disposable = this._disposables.pop();
      if (disposable) {
        disposable.dispose();
      }
    }
  }
}
